src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -27,17 +37,38 @@ private void InitSort(CultureInfo culture)
             }
             else
             {
-                Interop.Globalization.ResultCode resultCode = Interop.Globalization.GetSortHandle(GetNullTerminatedUtf8String(_sortName), out _sortHandle);
-                if (resultCode != Interop.Globalization.ResultCode.Success)
-                {
-                    Interop.Globalization.CloseSortHandle(_sortHandle);
+                _isAsciiEqualityOrdinal = (_sortName == "en-US" || _sortName == "");
 
-                    if (resultCode == Interop.Globalization.ResultCode.OutOfMemory)
-                        throw new OutOfMemoryException();
+                lock (_lock)
+                {
+                    if (s_sortNameToSortHandleCache == null)
+                    {
+                        s_sortNameToSortHandleCache = new Dictionary<string, IntPtr>();
+                    }
 
-                    throw new ExternalException(SR.Arg_ExternalException);
+                    if (!s_sortNameToSortHandleCache.TryGetValue(_sortName, out _sortHandle))
+                    {
+                        s_sortNameToSortHandleCache.Add(_sortName, (_sortHandle = GetSortHandle(_sortName)));
```

Comment:
It would be nice to release the handle if `s_sortNameToSortHandleCache.Add` fails to add it.

Reply:
I did not want to use `TryAdd` because I did not know which exception I should throw on failure so I just added a `catch` block for `Add`.